File: searchin2darray.py
# // Time Complexity : O(M+N)
# // Space Complexity : O(1)
# // Did this code successfully run on Leetcode : yes
# // Three line explanation of solution in plain english
"""
Starting at the bottom left, we have two choices. If our current element is greater than the target, we need to go up.
If our current element is less than the target, we then go to the right. This only works if we start from the bottom left.
"""
class Solution:
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        m = len(matrix)-1
        n = 0
        
        while(m >= 0 and n <= len(matrix[0]) - 1):
            if(matrix[m][n] > target):
                m -= 1
            elif(matrix[m][n] < target):
                n += 1
            else:
                return True
                
        return False
            
        
        
        

File: search_in_2d_array.java
//Time Complexity: O(m+n)

class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if(matrix == null || matrix.length == 0) return false;
        int m = matrix.length; int n = matrix[0].length;
        int r = 0; int c = n-1;
        
        while(r < m && c >= 0){
            if(matrix[r][c] == target) return true;
            else if(matrix[r][c] < target) r++;
            else c--;
        }
        return false;
        
        
    }
}
